lib/linked_list.rb

```diff
@@ -12,150 +11,321 @@ def initialize(value, next_node = nil)
-      raise NotImplementedError
+  # method to print all the values in the linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def visit
+    if @head.nil 
+      return nil 
+    else
+      current = @head 
+      until current.nil?
+        puts(current.data)
+        current = current.next 
+      end
     end
+  end
 
-    # method to reverse the singly linked list
-    # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def reverse
-      raise NotImplementedError
+  # method to delete the first node found with specified value
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def delete(value)
+    if @head.nil?
+      return
+    elsif @head.data == value
+      @head = @head.next
+    else
+      current = @head
+      until current.next.nil?
+        if current.next.data == value
+          current.next = current.next.next
+        else
+          current = current.next
+        end
+      end
     end
+  end
 
-
-    ## Advanced Exercises
-    # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_middle_value
-      raise NotImplementedError
+  # method to reverse the singly linked list
+  # note: the nodes should be moved and not just the values in the nodes
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def reverse
+    if @head.nil?
+      return nil
+    else
+      current = @head
+      previous = nil 
+      temp = current.next 
+      until current.next.nil?
+        temp = current.next 
+        current.next = previous
+        previous = current 
+        current = temp 
+      end
+      current.next = previous
+      @head = current
     end
+  end
 
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_nth_from_end(n)
-      raise NotImplementedError
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  # Time Complexity: O(2n) which is O(n)
+  # Space Complexity: O(1)
+  def find_middle_value
+    if @head.nil?
+      return nil 
+    else
+      array_length = self.length() #O(n)
+      if array_length % 2 == 0 
+        middle = (array_length / 2 ) - 1
+      else
+        middle = array_length / 2 
+      end
+      return get_at_index(middle) #O(n)
     end
+  end
 
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def has_cycle
-      raise NotImplementedError
+  # find the nth node from the end and return its value
+  # assume indexing starts at 0 while counting to n
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def find_nth_from_end(n)

```

Comment:
It might be easier to move `current` n spaces down.  Then set a variable called `n_back` to head and move `current` and `n_back` 1 step with each iteration of the loop until current is `nil`.